flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.flink.table.utils.EncodingUtils;
+
+import java.util.Map;
+
+/**
+ * Rowtime descriptor for describing an event time attribute in the schema.
+ */
+@PublicEvolving
+public class Rowtime implements Descriptor {
+
+	private final DescriptorProperties internalProperties = new DescriptorProperties(true);
+
+	/**
+	 * These constants will be removed once RowtimeValidator is ported to flink-table-common.
+	 */
+	private static final String ROWTIME_TIMESTAMPS_TYPE = "rowtime.timestamps.type";
+	private static final String ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD = "from-field";
+	private static final String ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_SOURCE = "from-source";
+	private static final String ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM = "custom";
+	private static final String ROWTIME_TIMESTAMPS_FROM = "rowtime.timestamps.from";
+	private static final String ROWTIME_TIMESTAMPS_CLASS = "rowtime.timestamps.class";
+	private static final String ROWTIME_TIMESTAMPS_SERIALIZED = "rowtime.timestamps.serialized";
+
+	private static final String ROWTIME_WATERMARKS_TYPE = "rowtime.watermarks.type";
+	private static final String ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_ASCENDING = "periodic-ascending";
+	private static final String ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_BOUNDED = "periodic-bounded";
+	private static final String ROWTIME_WATERMARKS_TYPE_VALUE_FROM_SOURCE = "from-source";
+	private static final String ROWTIME_WATERMARKS_TYPE_VALUE_CUSTOM = "custom";
+	private static final String ROWTIME_WATERMARKS_CLASS = "rowtime.watermarks.class";
+	private static final String ROWTIME_WATERMARKS_SERIALIZED = "rowtime.watermarks.serialized";
+	private static final String ROWTIME_WATERMARKS_DELAY = "rowtime.watermarks.delay";
+
+	/**
+	 * Sets a built-in timestamp extractor that converts an existing {@link Long} or
+	 * {@link Types#SQL_TIMESTAMP} field into the rowtime attribute.
+	 *
+	 * @param fieldName The field to convert into a rowtime attribute.
+	 */
+	public Rowtime timestampsFromField(String fieldName) {
+		internalProperties.putString(
+			ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD);
+		internalProperties.putString(ROWTIME_TIMESTAMPS_FROM, fieldName);
+		return this;
+	}
+
+	/**
+	 * Sets a built-in timestamp extractor that converts the assigned timestamps from
+	 * a DataStream API record into the rowtime attribute and thus preserves the assigned
+	 * timestamps from the source.
+	 *
+	 * <p>Note: This extractor only works in streaming environments.
+	 */
+	public Rowtime timestampsFromSource() {
+		internalProperties.putString(
+			ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_SOURCE);
+		return this;
+	}
+
+	/**
+	 * Sets a custom timestamp extractor to be used for the rowtime attribute.
+	 *
+	 * @param extractor The {@link TimestampExtractor} to extract the rowtime attribute
+	 *                  from the physical type.
+	 */
+	public Rowtime timestampsFromExtractor(TimestampExtractor extractor) {
```

Comment:
This changes the original logic. Before we were only serializing as the last option. Serializing is dangerous as we might not be able to restore the base64 string. Please reintroduce the original logic.